File: backend/backend/models/jobs.py
from sqlalchemy import Column, String, Boolean, Integer, DateTime, ForeignKey, Float, JSON
from sqlalchemy.sql import func
from backend.database.session import Base
import uuid


def generate_uuid():
    return str(uuid.uuid4())


class Job(Base):
    """A scheduled or queued job for background execution."""
    __tablename__ = "jobs"

    id = Column(String, primary_key=True, default=generate_uuid)
    title = Column(String, nullable=False)
    job_type = Column(String, nullable=False, index=True)  # orchestration, chat, tool, custom
    payload = Column(JSON, nullable=False)  # job-specific data
    priority = Column(Integer, nullable=False, default=5, index=True)  # 1=highest, 10=lowest
    status = Column(String, nullable=False, default="queued", index=True)  # queued, running, completed, failed, cancelled, paused
    progress = Column(Integer, nullable=False, default=0)  # 0-100
    result = Column(JSON, nullable=True)
    error_message = Column(String, nullable=True)
    retry_count = Column(Integer, nullable=False, default=0)
    max_retries = Column(Integer, nullable=False, default=3)
    conversation_id = Column(String, ForeignKey("conversations.id", ondelete="SET NULL"), nullable=True, index=True)
    session_id = Column(String, ForeignKey("orchestration_sessions.id", ondelete="SET NULL"), nullable=True, index=True)
    scheduled_at = Column(DateTime(timezone=True), nullable=True)  # when to run (null = immediately)
    started_at = Column(DateTime(timezone=True), nullable=True)
    completed_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())


class JobLog(Base):
    """Structured log entries for job execution."""
    __tablename__ = "job_logs"

    id = Column(String, primary_key=True, default=generate_uuid)
    job_id = Column(String, ForeignKey("jobs.id", ondelete="CASCADE"), nullable=False, index=True)
    level = Column(String, nullable=False, default="info")  # debug, info, warn, error
    message = Column(String, nullable=False)
    log_metadata = Column("metadata", JSON, nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())

File: app/src/renderer/src/lib/api/jobs.ts
import { apiClient } from "./client";

// ── Types ────────────────────────────────────────────────────

export type JobStatus = "queued" | "running" | "completed" | "failed" | "cancelled" | "paused";
export type JobType = "orchestration" | "chat" | "tool" | "custom";

export type JobRecord = {
  id: string;
  title: string;
  jobType: JobType;
  priority: number;
  status: JobStatus;
  progress: number;
  retryCount: number;
  maxRetries: number;
  errorMessage: string | null;
  startedAt: string | null;
  completedAt: string | null;
  createdAt: string | null;
};

export type JobDetail = JobRecord & {
  result: Record<string, any> | null;
  logs: JobLogRecord[];
};

export type JobLogRecord = {
  level: string;
  message: string;
  createdAt: string;
};

export type CreateJobPayload = {
  title: string;
  job_type: JobType;
  payload?: Record<string, any>;
  priority?: number;
  max_retries?: number;
  conversation_id?: string;
  session_id?: string;
};

// ── API ──────────────────────────────────────────────────────

export const jobsApi = {
  async create(payload: CreateJobPayload): Promise<{ id: string; title: string; jobType: string; priority: number; status: string; progress: number }> {
    return apiClient.post("/jobs", payload);
  },

  async list(params?: { status?: string; job_type?: string; limit?: number }): Promise<JobRecord[]> {
    const query = new URLSearchParams();
    if (params?.status) query.append("status", params.status);
    if (params?.job_type) query.append("job_type", params.job_type);
    if (params?.limit !== undefined) query.append("limit", String(params.limit));
    const qs = query.toString() ? `?${query.toString()}` : "";
    return apiClient.get<JobRecord[]>(`/jobs${qs}`);
  },

  async get(id: string): Promise<JobDetail> {
    return apiClient.get<JobDetail>(`/jobs/${id}`);
  },

  async cancel(id: string): Promise<{ id: string; status: string }> {
    return apiClient.post(`/jobs/${id}/cancel`);
  },

  async pause(id: string): Promise<{ id: string; status: string }> {
    return apiClient.post(`/jobs/${id}/pause`);
  },

  async resume(id: string): Promise<{ id: string; status: string }> {
    return apiClient.post(`/jobs/${id}/resume`);
  },
};
